src/status.ts

```diff
@@ -2,87 +2,139 @@
 /* eslint-disable @typescript-eslint/ban-types */
 import {
   emptyExecution,
+  EnforcedObjectProperties,
   Execution,
+  ExecutionSettings,
   Process,
   ProcessMessage,
+  Route,
+  Step,
   UpdateExecution,
 } from './types'
 import { deepClone } from './utils'
+/**
+ * Manages status updates of a route and provides various functions for tracking processes
+ * @param  {Route} route  The route this StatusManger belongs to.
+ * @param  {EnforcedObjectProperties<ExecutionSettings>} settings   The ExecutionSettings for this route
+ * @return {StatusManager}       An instance of StatusManager.
+ */
+export class StatusManager {
+  route: Route
+  settings: EnforcedObjectProperties<ExecutionSettings>
 
-export const initStatus = (
-  updateStatus?: (execution: Execution) => void,
-  initialStatus?: Execution
-) => {
-  const status = initialStatus || (deepClone(emptyExecution) as Execution)
-  // eslint-disable-next-line no-console
-  const update = updateStatus || console.log
-  if (!initialStatus) {
-    update(status)
+  constructor(
+    route: Route,
+    settings: EnforcedObjectProperties<ExecutionSettings>
+  ) {
+    this.route = route
+    this.settings = settings
   }
-  return { status, update }
-}
 
-export const createAndPushProcess = (
-  id: string,
-  updateStatus: (execution: Execution) => void,
-  status: Execution,
-  message: ProcessMessage,
-  params?: object
-): Process => {
-  const process = status.process.find((p) => p.id === id)
-  if (process) {
-    status.status = 'PENDING'
-    updateStatus(status)
-    return process
-  }
-  const newProcess: Process = {
-    id: id,
-    startedAt: Date.now(),
-    message: message,
-    status: 'PENDING',
-  }
-  if (params) {
-    for (const [key, value] of Object.entries(params)) {
-      newProcess[key] = value
+  /**
+   * Initializes the execution object of a Step.
+   * @param  {Step} step  The current step in execution
+   * @return {Execution, UpdateExecution}       The initialized execution object for this step and a function to update this step
+   */
+  initExecutionObject = (step: Step) => {
+    const currentExecution =
+      step.execution || (deepClone(emptyExecution) as Execution)
+    const updateExecution = (newExecution: Execution) => {
+      step.execution = newExecution
+      this.settings.updateCallback(this.route)
+    }
```

Comment:
This updates the step, but the route returned is still the old one. Does this only work because the step has a reference to the route? If yes we need to make this explicit. If a developer deep clones the step on his side this could break I guess.